File: src/main/java/com/tram/springbootangularboard/security/filter/FormLoginFilter.java
package com.tram.springbootangularboard.security.filter;

import com.tram.springbootangularboard.common.CommonUtils;
import com.tram.springbootangularboard.dto.FormLoginDto;
import com.tram.springbootangularboard.security.token.PreAuthorizationToken;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter;
import org.springframework.security.web.authentication.AuthenticationFailureHandler;
import org.springframework.security.web.authentication.AuthenticationSuccessHandler;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class FormLoginFilter extends AbstractAuthenticationProcessingFilter {

    private AuthenticationSuccessHandler authenticationSuccessHandler;
    private AuthenticationFailureHandler authenticationFailureHandler;
    public FormLoginFilter(String defaultFilterProcessesUrl) {
        super(defaultFilterProcessesUrl);
    }

    public FormLoginFilter(String defaultFilterProcessesUrl, AuthenticationSuccessHandler authenticationSuccessHandler, AuthenticationFailureHandler authenticationFailureHandler) {
        this(defaultFilterProcessesUrl);
        this.authenticationSuccessHandler = authenticationSuccessHandler;
        this.authenticationFailureHandler = authenticationFailureHandler;
    }

    //인증 시도 시 프로바이더를 통해 인증 정보 유효성 검증
    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException, IOException, ServletException {
        FormLoginDto formLoginDto = CommonUtils.getObjectMapper().readValue(request.getReader(), FormLoginDto.class);
        PreAuthorizationToken token = new PreAuthorizationToken(formLoginDto);
        //AuthenticationManager가 위의 PreAuthorizationToken 형태에 맞는 Provider를 찾아서 인증을 진행한다.
        return super.getAuthenticationManager().authenticate(token);
    }

    @Override
    protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication authResult) throws IOException, ServletException {
        this.authenticationSuccessHandler.onAuthenticationSuccess(request, response, authResult);
    }

    //TODO AuthenticationFailureHandler 직접 구현해서 적절한 응답을 리턴하도록 수정할 것
    @Override
    protected void unsuccessfulAuthentication(HttpServletRequest request, HttpServletResponse response, AuthenticationException failed) throws IOException, ServletException {
        this.authenticationFailureHandler.onAuthenticationFailure(request, response, failed);
    }
}

File: src/main/java/com/tram/springbootangularboard/security/filters/FormLoginFilter.kt
package com.tram.springbootangularboard.security.filters

import org.springframework.security.core.Authentication
import org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


class FormLoginFilter : AbstractAuthenticationProcessingFilter {
    constructor(defaultFilterProcessesUrl: String) : super(defaultFilterProcessesUrl)
    constructor(defaultFilterProcessesUrl: String, authenticationSuccessHandler: AuthenticationSuccessHandler, authenticationFailureHandler: AuthenticationFailureHandler) : this(defaultFilterProcessesUrl) {
        this.setAuthenticationSuccessHandler(authenticationSuccessHandler)
        this.setAuthenticationFailureHandler(authenticationFailureHandler)
    }
    override fun attemptAuthentication(request: HttpServletRequest?, response: HttpServletResponse?): Authentication {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}